tests/scanner/scanners/external_project_access_scanner_test.py

```diff
@@ -0,0 +1,301 @@
+from tests.unittest_utils import get_datafile_path
+# pylint: enable=line-too-long
+
+TEST_EMAILS = ['user1@example.com', 'user2@example.com']
+
+LOGGER = logger.get_logger(__name__)
+
+
+class ExternalProjectAccessScannerTest(ForsetiTestCase):
+    """Tests for ExternalProjectAccessScanner"""
+
+    def setUp(self):
+        """Setup"""
+        self.global_configs = dict()
+        self.scanner_configs = (
+            dict(output_path='gs://test-forseti-dev/scanner/output',
+                 rules_path=__file__,
+                 scanners=[dict(name='external_project_access', enabled=True)]))
+
+        mock_invnetory_config = mock.MagicMock()
+        mock_invnetory_config.get_root_resource_id.return_value = (
+            'organizations/567890')
+
+        mock_service_config = mock.MagicMock()
+        mock_service_config.get_inventory_config.return_value = (
+            mock_invnetory_config)
+
+        self.service_config = mock_service_config
+        self.model_name = 'TestModel'
+        self.snapshot_timestamp = datetime.now().strftime('%Y%m%dT%H%M%SZ')
+        self.rules = get_datafile_path(
+            __file__, 'external_project_access_test_rules_1.yaml')
+
+        self.stash_email_method = epas.get_user_emails
+
+    def test_mock_emails(self):
+        """Test obtaining e-mails from mock"""
+        epas.get_user_emails = mock.MagicMock(return_value=TEST_EMAILS)
+        self.assertEqual(len(epas.get_user_emails(None)), 2)
+
+    #pylint: disable=W0212
+    def test_retrieve(self):
+        """Test retrieving project ancestry data"""
+        list_to_generator = lambda x: (n for n in x)
+
+        epas.get_user_emails = mock.MagicMock(return_value=TEST_EMAILS)
+        epas.extract_project_ids = mock.MagicMock(
+            return_value=['project1', 'project2', 'project3'])
+
+        api_helpers.get_delegated_credential = mock.MagicMock()
+
+        crm.get_projects = mock.MagicMock(
+            return_value=list_to_generator(
+                [fake_crm_responses.FAKE_PROJECTS_API_RESPONSE1]))
+
+        crm.CloudResourceManagerClient.get_project_ancestry = mock.MagicMock(
+            return_value=json.loads(
+                fake_crm_responses.GET_PROJECT_ANCESTRY_RESPONSE)['ancestor'])
+
+        scanner = epas.ExternalProjectAccessScanner(self.global_configs,
+                                                    self.scanner_configs,
+                                                    self.service_config,
+                                                    self.model_name,
+                                                    self.snapshot_timestamp,
+                                                    self.rules)
+        scanner.rules_engine.build_rule_book(
+            self.service_config.get_inventory_config())
+
+        user_ancestries = scanner._retrieve()
+
+        self.assertIn('user1@example.com', user_ancestries.keys())
+        self.assertTrue(isinstance(user_ancestries['user1@example.com'], list))
```

Comment:
Instead of doing very detailed and repetitive asserts like this, can we not assert on the what the expected values are?
```
expected_result = [1, 2,3]
actual_result = scanner.retrieve()
self.assertTrue(expected_result, actual_result)
```

Reply:
the expected result is not a simple value, it is a nested array of user information and project instances. verifying the expected result by its value might involve creating those project instances, and matching the instance properties one by one to make sure the expected instance and the actual instance is equal by value.

Reply:
However, please feel free to suggest what would be the best way to test here. The expected result should look like this: 
{'user1@example.com': [[project<id=forseti-system-test,parent=None>, folder<id=3333333,parent=None>, organization<id=2222222,parent=None>], [project<id=forseti-system-test,parent=None>, folder<id=3333333,parent=None>, organization<id=2222222,parent=None>], [project<id=forseti-system-test,parent=None>, folder<id=3333333,parent=None>, organization<id=2222222,parent=None>]], 'user2@example.com': [[project<id=forseti-system-test,parent=None>, folder<id=3333333,parent=None>, organization<id=2222222,parent=None>], [project<id=forseti-system-test,parent=None>, folder<id=3333333,parent=None>, organization<id=2222222,parent=None>], [project<id=forseti-system-test,parent=None>, folder<id=3333333,parent=None>, organization<id=2222222,parent=None>]]}